### ISSUE TITLE
insertPackage is slow for cloud-based file shares

### ISSUE THREAD

**Issue by natbprice** [eddelbuettel/drat#128]
**Use-case**
1.  Package repo on cloud-based file share (See [SMB file shares in Azure files](https://docs.microsoft.com/en-us/azure/storage/files/files-smb-protocol?tabs=azure-portal))
    - Secure/low-cost storage
    - Accessible anywhere (on/off premise, Windows/Linux)
2. Large R packages 
    - No limits on package size (Git LFS + low-cost cloud storage)
    - e.g., pre-trained ML models, datasets, visualizations, etc.

**Issue**
- `insertPackage` is slow on cloud-based file share because it relies on base `file.copy`
- Base `file.copy` is slow on cloud-based file share because it performs a large number of write operations 
    - (MB/sec) = (MB/write) x (write/sec) ?
    - Excessive overhead due to large number of write operations results in very low throughput
- Number of `file.copy` write operations is high because C `BUFSIZ` is small (See [How big is your BUFSIZ?](https://c-for-dummies.com/blog/?p=4711) and Line 515 `src/main/platform.c`)
    - Windows - 512 bytes
    - Linux - 8192 bytes (16X)

**Possible Solutions**
1. Use `fs::file_copy` instead of base `file.copy` (Line 134 in insertPackage.R)
    - Consider if `ls` should be suggested package with fallback to base `file.copy`
    - `fs::file_copy` is faster because it performs fewer write operations
    - Buffer size is much larger than `file.copy` on any OS 
        - 65536 byes (128X) ? ([libuv documenation](http://docs.libuv.org/en/v1.x/handle.html#c.uv_alloc_cb))
    - Trivial change to `drat` and likely some performance benefits (regardless of OS)
2. Submit patch for base `file.copy`
    - Uncertain if patch would be accepted
    - Accepted buffer size may still be much smaller than `fs::file_copy` even if it is agreed to increase it
    - Performance may still be poor even if it is improved

```
Index: src/main/platform.c
===================================================================
--- src/main/platform.c	(revision 81244)
+++ src/main/platform.c	(working copy)
@@ -515,13 +515,13 @@
  *  the second set of files to be appended to the first.
  */
 
-#if defined(BUFSIZ) && (BUFSIZ > 512)
+#if defined(BUFSIZ) && (BUFSIZ > 65536)
 /* OS's buffer size in stdio.h, probably.
    Windows has 512, Solaris 1024, glibc 8192
  */
 # define APPENDBUFSIZE BUFSIZ
 #else
-# define APPENDBUFSIZE 512
+# define APPENDBUFSIZE 65536
 #endif
 
 static int R_AppendFile(SEXP file1, SEXP file2)
```

**Comment by eddelbuettel** [comment]
Thanks for the detailed report.  However, as this repo is neither base R nor fs so filing it here makes little sense overall.  

Could I ask you to maybe now stop with your little vendatta _at this repo_?  Can we agree that two tickets is now sufficient? 

Maybe write some blog posts about it or do whatever it takes to get this out of your system but this is redundant here.  If you go back to earlier issue tickets you will see that windows performance (or the lack thereof) has been discussed.  But this is an _application_ level package and _not_ the place to change this.  Tomas, who is the R Core member who looks after r-devel is a reasonable person so maybe you discuss with him, and/or start on r-devel.  Increasing the buffer size _may_ in fact be the right call especially as R 4.2.0 will, if I remember correctly, leave 32-bit windows behind.  

I will close this now.

**Comment by natbprice** [comment]
Hey Dirk,

I think we got off on the wrong foot and there has been some misunderstanding/miscommunication. I tried to clearly describe the issue to the best of my ability and performed additional research to support my request as documented in this issue. This ticket is meant to help improve drat; I definitely don't have a vendetta against drat.

The solution I proposed is to drat so I thought it was appropriate to post it here. In my mind it is a relatively simple change (3 lines?): use fs::file_copy if it is already installed, but don't require it as a dependency.

I agree that any changes to R core is a separate issue. You suggested changing R core so I discussed that solution and included a patch file. I am not sure that is the best solution (or a feasible solution at all) as I described in issue. I am currently unable to submit bug reports to R, but I have emailed to request access.

I respect your decision to not add any additional dependencies in drat. I hope you don't mind if I fork the repository to make it work for my use case.

Best,

Nat
